pyro/distributions/empirical.py

```diff
@@ -0,0 +1,160 @@
+            log_weight = torch.tensor(log_weight).type(weight_type)
+
+        # Seed the container tensors with the correct tensor types
+        if self._samples is None:
+            self._samples = value.new()
+            self._log_weights = log_weight.new()
+        # Append to the buffer list
+        self._samples_buffer.append(value)
+        self._weights_buffer.append(log_weight)
+
+    def sample(self, sample_shape=torch.Size()):
+        self._finalize()
+        idxs = self._categorical.sample(sample_shape=sample_shape)
+        return self._samples[idxs]
+
+    def log_prob(self, value):
+        """
+        Returns the log of the probability mass function evaluated at ``value``.
+        Note that this currently only supports scoring single values and not
+        a tensor of values.
+
+        :param torch.Tensor value: scalar or tensor value to be scored.
+        """
+        if is_validation_enabled():
+            if value.size() != self.event_shape:
+                raise ValueError("``value.size()`` must be {}".format(self.event_shape))
+        self._finalize()
+        selection_mask = self._samples.eq(value).contiguous().view(self.sample_size, -1)
+        # Return -Inf if value is outside the support.
+        if not selection_mask.any():
+            return self._log_weights.new_zeros(torch.Size()).log()
+        idxs = torch.arange(self.sample_size)[selection_mask.min(dim=-1)[0]]
+        log_probs = self._categorical.log_prob(idxs)
+        return log_sum_exp(log_probs)
+
+    @property
+    def batch_shape(self):
+        return torch.Size()
+
+    @property
+    def event_shape(self):
+        self._finalize()
+        if self._samples is None:
+            return None
+        return self._samples.size()[1:]
+
+    @property
+    def mean(self):
+        self._finalize()
+        if self._samples.dtype in (torch.int32, torch.int64):
+            raise ValueError("Mean for discrete empirical distribution undefined. Consider converting samples " +
+                             "to ``torch.float32`` or ``torch.float64``.")
+        weights = self._log_weights
+        for _ in range(self._samples.dim() - 1):
+            weights = weights.unsqueeze(-1)
+        return (log_sum_exp(weights, scale=self._samples, dim=0) - log_sum_exp(weights, dim=0)).exp()
+
+    @property
+    def variance(self):
+        self._finalize()
+        if self._samples.dtype in (torch.int32, torch.int64):
+            raise ValueError("Variance for discrete empirical distribution undefined. Consider converting samples " +
+                             "to ``torch.float32`` or ``torch.float64``.")
+        deviation_squared = torch.pow(self._samples - self.mean, 2)
+        weights = self._log_weights
+        for _ in range(self._samples.dim() - 1):
+            weights = weights.unsqueeze(-1)
+        return (log_sum_exp(weights, scale=deviation_squared, dim=0) - log_sum_exp(weights, dim=0)).exp()
```

Comment:
ditto: it seems like it would be clearer to
```py
weights = (self._log_weights - log_sum_exp(self._log_weights)).exp()
weights = weights.reshape(weights.shape + (1,) * self.event_dim)
mean = (weights * self._samples).sum(0)
return (weights * (self._samples - mean).pow(2)).sum(0)
```

Reply:
I had it that way to begin with. Should I just move over the `log_sum_exp` logic here if the scale factor seems confusing?

Reply:
It's up to you. I didn't realize the scale arg was provided by scipy.